### PR TITLE
Properties view for the vscode extension

### PR DESCRIPTION
**Pull request by ogoffart**
Some refactoring will be required later because the current interface is a bit awkward, especially the `BindingTextProvider` because the webview can't easily access the text editor as it can only communicate to the extension with messages.
But this get the thing done.

### CODE REVIEW

File: editors/vscode/src/propertiesView.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        const line_utf8 = new TextEncoder().encode(this.#code[l]);
+        let l2 = location.expression_range.end.line;
+        if (l == l2) {
+            return new TextDecoder().decode(
+                line_utf8.slice(location.expression_range.start.character, location.expression_range.end.character));
+        }
+        let result = new TextDecoder().decode(
+            line_utf8.slice(location.expression_range.start.character));
+        l++;
+        while (l < l2) {
```
Comment (tronical) [inline editors/vscode/src/propertiesView.ts:32]: My first thought when looking at this comparison was: "Why is it `l < 12`? Why 12?" 🤣

I suggest to rename l2 to something else :). Maybe `last_line`?

File: editors/vscode/src/common.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        });
+    }
+
+    private _getHtmlForWebview(webview: vscode.Webview) {
+
+        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'out/propertiesView.js'));
+        const nonce = getNonce();
+
+
+        // FIXME: share with the online editor?
```
Comment (tronical) [inline editors/vscode/src/common.ts:101]: Not part of this patch, but just thinking out loud: Perhaps we can use a bundler to bundle `colors.css` (inline), the shared `properties.ts`, `propertiesView.ts` and a `.html` template (as inline below) into one file and also get rid of the nonce.

**Review by tronical — COMMENTED** [review]
Looks generally good to me. I'll let Tobias make the call on the online editor bits :)

File: editors/vscode/src/common.ts (hunk 2 above)
Comment (ogoffart) [reply editors/vscode/src/common.ts:97]: Ideally, we should somehow share the .css between the online editor and this.

File: tools/online_editor/src/editor_widget.ts [hunk 3]
```diff
@@ -4,12 +4,14 @@
 // cSpell: ignore lumino inmemory mimetypes printerdemo
 
 import { slint_language } from "./highlighting";
+import {
+  lsp_range_to_editor_range,
+} from "./lsp_integration";
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply tools/online_editor/src/editor_widget.ts:9]: VSCode has its own range type, and also its own conversion function from LSP to its range.

File: tools/online_editor/src/shared/properties.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+export interface PropertyQuery {
+  source_uri: string;
+  element: Element | null;
+  properties: Property[];
+}
+
+const TYPE_PATTERN = /^[a-z-]+$/i;
+
+let FOCUSING = false;
```
[the remark this answers is not in the document]
Comment (ogoffart) [reply tools/online_editor/src/shared/properties.ts:45]: I don't think so.  But maybe.

**Comment by ogoffart** [comment]
> Should we maybe make the LSP report the expression used in a property as a string?

That would solve the problem.
Another way would be to process all the string before giving that to the view.
